onnxruntime/core/providers/cuda/reduction/reduction_functions.cc

```diff
@@ -105,8 +105,42 @@ ApplicableMatrixReduction get_applicable_matrix_reduction(
     return ApplicableMatrixReduction::None;
   }
 
-  const auto rank = gsl::narrow<int64_t>(dims.size());
-  const auto min_and_max_axes = GetMinAndMaxContiguousAxes(rank, dims, original_axes);
+
+  // Remove all dims with value 1. This can help to optimize case like:
+  // dims=[2,3,1,4,1,5] and axes=[0,2,4], which is same as dims=[2,3,4,5] and axes=[0].
+  std::vector<int64_t> new_dims;
+  std::vector<int64_t> new_axes;
+  const auto original_rank = gsl::narrow<int64_t>(dims.size());
+  std::set<int64_t> original_axes_set;
+  for (const auto axis : original_axes) {
+    original_axes_set.insert(HandleNegativeAxis(axis, original_rank));
+  }
+
+  int64_t new_axis = 0;
+  for (size_t i = 0; i < dims.size(); i++) {
+    if (dims[i] != 1) {
+      new_dims.emplace_back(dims[i]);
+      if (original_axes_set.find(gsl::narrow<int64_t>(i)) != original_axes_set.end()) {
+        new_axes.emplace_back(new_axis);
+      }
+      new_axis++;
+    }
+  }
+
+  // Empty axes means reduce all dimensions, which has different meaning,
+  // so add a new dim to the end if all original axes are on dims with value 1.
+  if (!original_axes.empty() && new_axes.empty()) {
+    new_dims.emplace_back(1);
+    new_axes.emplace_back(new_axis);
+  }
+
+  // If all dims are value 1, make sure it's not empty by adding a new dim.
```

Comment:
this should also result in an no-op, also need a memcpy from input to output for this case.

Reply:
Same as above with different shapes: shape=[1,1], and axes=[], the new shape=[1] and new axes=[], the compute will work as normal.